### PR TITLE
Add HyperDualNumbersExt

### PR DESCRIPTION
**Pull request by CheukHinHoJerry**
Ref https://github.com/JuliaLinearAlgebra/Octavian.jl/issues/178#issue-1758088520

Basically I just replace the type constrain of 'ForwardDiff.Dual' to 'HyperDualNumbers.Hyper' and create an extra module. Haven't done any serious test on it but it works okay on my end.

I guess it might be possible to combine into a single ext with ForwardDiff.Dual but I will just leave something workable for now since I am not sure where/how should I do this properly.

### CODE REVIEW

**Comment by chriselrod** [comment]
Mind adding tests, adding hyper duals as a test dependency?
You can compare the answers with a naive or base multiplication.
I'd reinterpret to `Float64` before comparing, because comparisons tend to only check the value, and not any of the partials.

File: test/hyperduals.jl [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+  A2 = deepcopy(A1)
+  B2 = deepcopy(B1)
+  C2 = deepcopy(C1)
+
+  α = Float64(2.0)
+  β = Float64(2.0)
+
+  Octavian.matmul!(C1, A1, B1, α, β)
+  LinearAlgebra.mul!(C2, A2, B2, α, β)
+  @test C1 ≈ C2
```
Comment (chriselrod) [inline test/hyperduals.jl:32]:
```suggestion
  @test reinterpretH(C1) ≈ reinterpretHD(C2)
```

File: test/hyperduals.jl [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+  _x = zeros(HyperDualNumbers.Hyper{Float64}, size(x)...)
+  for i in eachindex(x)
+    _x = HyperDualNumbers.Hyper(x[i], rand(), rand(), rand())
+  end
+  return _x
+end
+
+function reinterpretHD(T, A)
+  tmp = reinterpret(T, A)
+  return tmp[1:4:end, :]
```
Comment (chriselrod) [inline test/hyperduals.jl:11]: why the slice?

Comment (CheukHinHoJerry) [reply test/hyperduals.jl:11]: reinterpreting a $N \times M$ matrix containing HyperDuals returns a $4N \times M$ matrix containing val, \eps1, \eps2 and \eps12 part of each of the entries. 

This is essentially checking only the val part of the matrix A.

I will do some updates and name the test properly on what they are checking.

Comment (chriselrod) [reply test/hyperduals.jl:11]: Don't we also want to check the epsilons?

Comment (CheukHinHoJerry) [reply test/hyperduals.jl:11]: I thought what you mean in https://github.com/JuliaLinearAlgebra/Octavian.jl/pull/179#issuecomment-1594156765 is that we only check val part. I will fix it now.

**Comment by CheukHinHoJerry** [comment]
Stucked on a bug on two dual array since I was not notcing it would be a bit difference as in ForwardDiff.Dual but just got that fix and it should be fine now. Basically I checked:

- real array mul from right
- real array mul from left
- tranpose arrays
- two dual arrays

The reinterpredHD function is now a sanity check which I uses LinearAlgebra.mul to calcualte only the val part of the result. I think this can be removed but it doesn't harm if we keep that.

File: test/hyperduals.jl (hunk 2 above)
Comment (chriselrod) [reply test/hyperduals.jl:11]: I meant that is what `isapprox` does, so we need to reinterpret to check them all.

Comment (CheukHinHoJerry) [reply test/hyperduals.jl:11]: I now check the entire thing and it should be ready for review. Sorry for iterating so many times.

File: test/hyperduals.jl [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+  C1 = rand(Float64, m, n)
+
+  α = Float64(2.0)
+  β = Float64(2.0)
+
+  @testset "real array from the right" begin
+    A1dual = randdual(A1)
+    C1dual = randdual(C1)
+    
+    A2dual = deepcopy(A1dual)
```
Comment (chriselrod) [inline test/hyperduals.jl:30]: Not a big deal, but some of these copies are unnecessary.

**Review by chriselrod — APPROVED** [review]
Okay, great!
One last thing: mind bumping the version, so we can tag a new release?

**Comment by CheukHinHoJerry** [comment]
> Okay, great! One last thing: mind bumping the version, so we can tag a new release?

Just bumped to 0.3.23. Thank you for your review.

**Pull request merged**